### PR TITLE
Add pkcs11 support to get_security_files

### PR DESCRIPTION
**Pull request by MiguelCompany**
This implements ros2/design#332

### CODE REVIEW

**Review by ruffsl — APPROVED** [review]

**Comment by ruffsl** [comment]
Ping @clalancette or @mikaelarguedas ? Good to merge?

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2023-02-16/29927/1

**Comment by SubaruArai** [comment]
@MiguelCompany any update on this pr?

Or should we wait [REP-2015](https://github.com/ros-infrastructure/rep/pull/375) to be approved first?

**Comment by MiguelCompany** [comment]
@SubaruArai I don't know what the process is. I guess we should wait for [REP-2015](https://github.com/ros-infrastructure/rep/pull/375), but I am just an external contributor, and I suppose that someone in the ROS 2 team should answer that.

Maybe @clalancette or @gbiggs can clarify?

**Review by ahcorde — CHANGES_REQUESTED** [review]
This PR is stuck for a while, are we planning to merge this before feature freeze? @clalancette ?

 I'm missing some tests in the `test_security.cpp` file. Do you mind to add them?

**Comment by MiguelCompany** [comment]
@ahcorde I just rebased this. Will try to add tests in the `test_security.cpp` file

**Comment by MiguelCompany** [comment]
@ahcorde I updated the tests in `test_security.cpp`

File: rmw_dds_common/test/test_security.cpp [hunk 1]
```diff
@@ -22,7 +22,33 @@
 
 #include "rmw_dds_common/security.hpp"
 
-TEST(test_security, files_exist_no_prefix)
+// Utility to write test content on required files
+template<size_t N>
```
Comment (ahcorde) [inline rmw_dds_common/test/test_security.cpp:26]: include `<cstddef>`

File: rmw_dds_common/src/security.cpp [hunk 2]
```diff
@@ -13,41 +13,128 @@
...
+  using std::placeholders::_2;
+  using std::placeholders::_3;
+  using std::placeholders::_4;
+  using security_file_processor =
+    std::function<bool (bool, const std::string &, const std::filesystem::path &, std::string &)>;
+  using processor_vector =
+    std::vector<std::pair<std::string, security_file_processor>>;
+
+
+  // Key: the security attribute
```
Comment (ahcorde) [inline rmw_dds_common/src/security.cpp:93]:
```suggestion
    std::vector<std::pair<std::string, security_file_processor>>;

  // Key: the security attribute
```

File: rmw_dds_common/src/security.cpp [hunk 3]
```diff
@@ -13,41 +13,128 @@
...
 
   const std::unordered_map<std::string, std::string> optional_files{
     {"CRL", "crl.pem"},
   };
 
-  for (const std::pair<const std::string, std::string> & el : required_files) {
-    std::filesystem::path full_path(secure_root);
-    full_path /= el.second;
-    if (!std::filesystem::is_regular_file(full_path)) {
+  for (const std::pair<const std::string,
```
Comment (ahcorde) [inline rmw_dds_common/src/security.cpp:120]: #include `<utility>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: rmw_dds_common/test/test_security.cpp (hunk 1 above)
Comment (MiguelCompany) [reply rmw_dds_common/test/test_security.cpp:27]: 305e4cc88ca6b8f59705c4c445d609ede74ab0fb

File: rmw_dds_common/src/security.cpp (hunk 2 above)
Comment (MiguelCompany) [reply rmw_dds_common/src/security.cpp]: 305e4cc88ca6b8f59705c4c445d609ede74ab0fb

File: rmw_dds_common/src/security.cpp (hunk 3 above)
Comment (MiguelCompany) [reply rmw_dds_common/src/security.cpp:119]: Was already there (line 19)

**Comment by clalancette** [comment]
Here is CI for this:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20766/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15064/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21447/)

**Comment by clalancette** [comment]
A couple more pieces of CI, then I think this is good to go:

* RHEL [Build Status](http://ci.ros2.org/job/ci_linux-rhel/630/)
* Clang [Build Status](http://ci.ros2.org/job/ci_linux_clang_libcxx/67/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21462/)

**Review by clalancette — APPROVED** [review]

**Review by ahcorde — APPROVED** [review]

**Comment by clalancette** [comment]
All of the warnings here were also in the nightlies, so going ahead and merging this one in.

**Pull request merged**